pkg/controller/database/cosmosdb/managed.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/crossplane/provider-azure/apis/v1alpha3"
+	"github.com/crossplane/provider-azure/pkg/clients/database/cosmosdb"
+)
+
+// Error strings
+const (
+	errProviderSecretNil  = "provider does not have a secret reference"
+	errNotNoSQLAccount    = "managed resource is not a Database Account"
+	errCreateNoSQLAccount = "cannot create Database Account"
+	errGetNoSQLAccount    = "cannot get Database Account"
+	errDeleteNoSQLAccount = "cannot delete Database Account"
+)
+
+// Setup adds a controller that reconciles NoSQLAccount.
+func Setup(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(v1beta1.CosmosDBAccountGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1beta1.CosmosDBAccount{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.CosmosDBAccountGroupVersionKind),
+			managed.WithConnectionPublishers(),
+			managed.WithExternalConnecter(&connecter{kube: mgr.GetClient(), newClientFn: cosmosdb.NewDatabaseAccountClient}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connecter struct {
+	kube        client.Client
+	newClientFn func(creds []byte) (cosmosdb.AccountClient, error)
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	r, ok := mg.(*v1beta1.CosmosDBAccount)
+	if !ok {
+		return nil, errors.New(errNotNoSQLAccount)
+	}
+
+	p := &v1alpha3.Provider{}
+	n := meta.NamespacedNameOf(r.Spec.ProviderReference)
+	if err := c.kube.Get(ctx, n, p); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider %s", n)
+	}
+
+	if p.GetCredentialsSecretReference() == nil {
+		return nil, errors.New(errProviderSecretNil)
+	}
+
+	s := &corev1.Secret{}
+	n = types.NamespacedName{Namespace: p.Spec.CredentialsSecretRef.Namespace, Name: p.Spec.CredentialsSecretRef.Name}
+	if err := c.kube.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get provider secret %s", n)
+	}
+
+	client, err := c.newClientFn(s.Data[p.Spec.CredentialsSecretRef.Key])
+	return &external{client: client}, errors.Wrap(err, "cannot create new Azure Database Account client")
+}
+
+// external is a createsyncdeleter using the Azure API.
+type external struct {
+	client cosmosdb.AccountClient
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	r, ok := mg.(*v1beta1.CosmosDBAccount)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotNoSQLAccount)
+	}
+
+	res, err := e.client.CheckNameExists(ctx, meta.GetExternalName(r))
+	if err != nil {
+		return managed.ExternalObservation{}, errors.Wrap(err, errGetNoSQLAccount)
+	}
+
+	if res.Response.StatusCode == http.StatusNotFound {
+		return managed.ExternalObservation{ResourceExists: false}, nil
+	}
```

Comment:
Could we use our existing [`azure.IsNotFound`](https://github.com/crossplane/provider-azure/blob/fabcb57ed9983fc1fdfe5dc906e79b01a66c1a04/pkg/clients/azure.go#L154) here? 

This could abbreviate the error case to:

```go
	res, err := e.client.CheckNameExists(ctx, meta.GetExternalName(r))
	if err != nil {
		return managed.ExternalObservation{}, errors.Wrap(resource.Ignore(azure.IsNotFound, err), errGetNoSQLAccount)
	}
```

Reply:
I tried it and just checked that it does not work. The behavior expected for this to work does not seem consistent across the Azure APIs.